shopfloor/services/zone_picking.py

```diff
@@ -646,15 +682,18 @@ def _scan_source_product(
         if not product:
             return response, message
         move_lines = self._find_location_move_lines(
-            locations=sublocation, product=product
+            locations=sublocation,
+            product=product,
+            package=package,
+            enforce_empty_package=self.work.menu.scan_location_or_pack_first,
         )
```

Comment:
For raising the right error message, we still need to know if there are lines with a package in case the location has been scanned but not the package.
```
move_lines = self._find_location_move_lines(
            locations=sublocation, product=product
            locations=sublocation,
            product=product,
            package=package,
        )
if not package and self.work.menu.scan_location_or_pack_first:
   move_lines_with_package_ids = []
   move_lines_without_package_ids = []
   for move_line in move_lines:
      if move_line.package_id:
          move_lines_with_package_ids.append(move_line.id)
      else:
          move_lines_without_package_ids.append(move_line.id)
    move_lines = move_lines.browse(move_lines_without_package_ids)
```
And then at the end you can check move_lines_with_package_ids for raising the right error message

Reply:
But before the end the default implementation of this scenario will process the first available move line (line 698..699) the end message you mention will never be displayed, no ?

Reply:
it will be if move_lines (i.e. without package) is empty

Reply:
Ok, but another issue with this change which prevents it to work. We have added the `enforce_empty_package` in the `find_location_move_lines` to use with the option scan location or package first.
Which will add the package into the search domain even if it is empty. So we will have no lines without package returned.

Reply:
that's why I dropped enforce_empty_package in the call